Cheek Samples

cassidysymons opened the pull request:
-- Description to be provided --

AmandaBirmingham commented on microsetta_private_api/repo/metadata_repo/_constants.py:
```diff
@@ -356,7 +356,18 @@
+        'env_package': 'human-skin',
+        'description': 'Microsetta Initiative Cheek sample',
+        'host_body_site': 'UBERON:skin of head'}
 }
```
In October, I had to check the constant metadata values for human cheek with Gail Ackermann, who is the final arbiter of these values in the Knight lab; she indicated that current best practice (it is constantly changing!) is that `host_body_site` should be `UBERON: cheek`, `sample_type` should be `skin of cheek`, and `env_material` should be `sebum material`.  Where does CMI get its guidance from?  I'm not surprised there are differences, but I'd be happier if we could align them ...

Also, for some point in the future, it would be nice to get these constants updated to v2 of EMPO, which specifies 4 levels instead of 3 (and it isn't just a bolt-on of one more level ... the whole level structure changes :-/ )

cassidysymons replied:
The values here are based on what was used for skin samples when AGP was actively collecting/processing them. We'll update these (cheek and other types) to reflect current recommendations.

AmandaBirmingham reviewed (commented):
One question/future request re constant metadata values for cheek samples.

AmandaBirmingham commented on microsetta_private_api/repo/sample_repo.py:
```diff
@@ -321,6 +330,106 @@ def update_info(self, account_id, source_id, sample_info,
+        # NB: We run this even if the barcode_meta dict is empty, as that
+        # means the answers associated with the last sample update should be
+        # purged. This applies both to the normal sample update process, as
+        # well as dissociate_sample().
```
This seems a little weird.  If this metadata is *truly* barcode-level (that is, sample level) and NOT source level; and if it is NOT survey data, then it doesn't make sense that it is wiped when a sample is dissociated from its source (and thus also its surveys)--you'd think that the sample-level metadata would remain, since the sample record itself continues to exist (just not attached to a source or survey).  

The fact that this metadata *should* be wiped on the dissociation of a sample from its source & survey(s) suggests that our modeling of this metadata as as "barcode-level" is flawed.  

At this point, I am not going to ask that we revisit this (I don't think anyone is in love with the current table solution, so I suspect we will be circling back to it sometime soon anyway) but I do think it is a code smell that we will eventually have to pay attention to.

cassidysymons replied:
The reason I treated it this way is because the dissociate_sample() function removes sample-related info such as collection date and sample site. From that perspective, I think that removing the values for these cheek fields is the logical move.

AmandaBirmingham commented on microsetta_private_api/repo/sample_repo.py:
```diff
@@ -451,3 +560,62 @@ def scrub(self, account_id, source_id, sample_id):
+                try:
+                    ret_val = datetime.datetime.strptime(
+                        fv,
+                        "%m/%d/%Y"
```
This hardcoded value is used twice in this file (here and in `_get_barcode_meta`), which to me means it should be stored in a string variable (well, constant) so that it can be found/changed/fixed if necessary.  

Same goes for the time format (which, I must note, is slightly *different* here than it is in `_get_barcode_meta`: "%I:%M %p" vs "%-I:%M %p" !)

cassidysymons replied:
It's not entirely logical on the surface, but `strptime()` and `strftime()` treat `%I` and `%-I` slightly differently. I created constants for all three formats (one date and two time) and included a comment explaining the reasoning for the date format differences.

AmandaBirmingham replied:
Thanks!

AmandaBirmingham commented on microsetta_private_api/db/patches/0144.sql:
```diff
@@ -0,0 +1,23 @@
+    ag_kit_barcode_id UUID NOT NULL PRIMARY KEY,
+    sample_site_last_washed_date DATE,
+    sample_site_last_washed_time TIME,
+    sample_site_last_washed_product SAMPLE_SITE_LAST_WASHED_PRODUCT_TYPE,
```
Is this allowed to be null?  Under what circumstances would that be the case?

AmandaBirmingham replied:
Ok ... now I'm wondering what the meaningful difference is between (a) a completed cheek sample with no record in the `ag_kit_barcodes_cheek` table and (b) a completed cheek sample with a record in the `ag_kit_barcodes_cheek` table for which all fields except the foreign key back to the `ag_kit_barcodes` table are NULL.  Do these two cases have two distinct meanings?  What is signified by a completely empty (except for FK) record in the cheek table?

cassidysymons replied:
From a metadata perspective, they do have a distinct meaning. NULL values would be reported as "not provided" while a cheek sample with no values for the questions (including NULL) would be reported as "not collected."

AmandaBirmingham replied:
Huh, that's fascinating--I don't think I really understand what these differences mean in this context, but good to know!

AmandaBirmingham commented on microsetta_private_api/model/sample.py:
```diff
@@ -74,18 +78,22 @@ def to_api(self):
 class SampleInfo:
-    def __init__(self, sample_id, datetime_collected, site, notes):
+    def __init__(self, sample_id, datetime_collected, site, notes,
+                 barcode_meta=None):
```
Curious: why does barcode_meta here default to `None` whereas in `Sample` it defaults to an empty dictionary?

AmandaBirmingham reviewed (changes requested):
Nothing major

cassidysymons commented:
@AmandaBirmingham I think this newest commit addresses the feedback that required changes, and I've added comments to notes that either don't seem to require change or were a strictly informational question. Please let me know if I missed anything or you have any further questions or concerns. Thanks!

wasade commented on microsetta_private_api/model/sample.py:
```diff
@@ -87,7 +87,7 @@ def to_api(self):
 class SampleInfo:
     def __init__(self, sample_id, datetime_collected, site, notes,
-                 barcode_meta=None):
+                 barcode_meta={}):
```
Can this be reverted? There is an important unexpected edgecase with default objects:

```
>>> def foo(a, b, example={}):
...   example[a] = b
...   print(example)
... 
>>> foo(1, 2)
{1: 2}
>>> foo(3, 4)
{1: 2, 3: 4}
>>> foo(1, 10)
{1: 10, 3: 4}
```

In almost all cases, the intended behavior and recommended approach is:

```
>>> def foo(a, b, example=None):
...   if example is None:
...     example = {}
...   example[a] = b
...   print(example)
... 
>>> foo(1, 2)
{1: 2}
>>> foo(3, 4)
{3: 4}
>>> foo(1, 10)
{1: 10}
```

These _should_ be caught by linters (see e.g. https://stackoverflow.com/a/73765790), which if not flagged here suggests we may need to revisit the linter configuration or linter used.

cassidysymons replied:
Done - https://github.com/biocore/microsetta-private-api/pull/597/commits/6a31b02cc7d4b9500c6a8800a1aab1de99c3de4e

wasade replied:
Thanks!
